security/advancedtls/advancedtls_integration_test.go

```diff
@@ -0,0 +1,509 @@
+	port    = ":50051"
+)
+
+// stageInfo contains a stage number indicating the current phase of each integration test, and a mutex.
+// Based on the stage number of current test, we will use different certificates and server authorization
+// functions to check if our tests behave as expected.
+type stageInfo struct {
+	mutex sync.Mutex
+	stage int
+}
+
+func (s *stageInfo) increase() {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	s.stage = s.stage + 1
+}
+
+func (s *stageInfo) read() int {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	return s.stage
+}
+
+// certStore contains all the certificates used in the integration tests.
+type certStore struct {
+	// clientPeer1 is the certificate sent by client to prove its identity. It is trusted by serverTrust1.
+	clientPeer1 tls.Certificate
+	// clientPeer2 is the certificate sent by client to prove its identity. It is trusted by serverTrust2.
+	clientPeer2 tls.Certificate
+	// serverPeer1 is the certificate sent by server to prove its identity. It is trusted by clientTrust1.
+	serverPeer1 tls.Certificate
+	// serverPeer2 is the certificate sent by server to prove its identity. It is trusted by clientTrust2.
+	serverPeer2  tls.Certificate
+	clientTrust1 *x509.CertPool
+	clientTrust2 *x509.CertPool
+	serverTrust1 *x509.CertPool
+	serverTrust2 *x509.CertPool
+}
+
+// loadCerts function is used to load test certificates at the beginning of each integration test.
+func (cs *certStore) loadCerts() error {
+	var err error
+	cs.clientPeer1, err = tls.LoadX509KeyPair(testdata.Path("client_cert_1.pem"),
+		testdata.Path("client_key_1.pem"))
+	if err != nil {
+		return err
+	}
+	cs.clientPeer2, err = tls.LoadX509KeyPair(testdata.Path("client_cert_2.pem"),
+		testdata.Path("client_key_2.pem"))
+	if err != nil {
+		return err
+	}
+	cs.serverPeer1, err = tls.LoadX509KeyPair(testdata.Path("server_cert_1.pem"),
+		testdata.Path("server_key_1.pem"))
+	if err != nil {
+		return err
+	}
+	cs.serverPeer2, err = tls.LoadX509KeyPair(testdata.Path("server_cert_2.pem"),
+		testdata.Path("server_key_2.pem"))
+	if err != nil {
+		return err
+	}
+	cs.clientTrust1, err = readTrustCert(testdata.Path("client_trust_cert_1.pem"))
+	if err != nil {
+		return err
+	}
+	cs.clientTrust2, err = readTrustCert(testdata.Path("client_trust_cert_2.pem"))
+	if err != nil {
+		return err
+	}
+	cs.serverTrust1, err = readTrustCert(testdata.Path("server_trust_cert_1.pem"))
+	if err != nil {
+		return err
+	}
+	cs.serverTrust2, err = readTrustCert(testdata.Path("server_trust_cert_2.pem"))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// serverImpl is used to implement pb.GreeterServer.
+type serverImpl struct{}
+
+// SayHello is a simple implementation of pb.GreeterServer.
+func (s *serverImpl) SayHello(ctx context.Context, in *pb.HelloRequest) (*pb.HelloReply, error) {
+	return &pb.HelloReply{Message: "Hello " + in.Name}, nil
+}
+
+func callAndVerifyInSameRPC(msg string, client pb.GreeterClient, shouldFail bool) error {
```

Comment:
I think the function name is a bit confusing, specifically "SameRPC" and "NewRPC" below. Are you referring to `ClientConn`? If that's the case then maybe we can rename them to `callAndVerifyWithClientConn` (this one), and `callAndVerify` the function below.